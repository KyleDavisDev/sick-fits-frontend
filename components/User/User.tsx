import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { ReactElement } from "react";

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        created
        id
        items {
          id
          quantity
          item {
            id
            price
            image
            title
            description
          }
        }
      }
    }
  }
`;

export interface IUserProps {
  children(data: any): ReactElement;
}

const User: React.FunctionComponent<IUserProps> = props => {
  return (
    <Query {...props} query={CURRENT_USER_QUERY}>
      {payload => {
        return props.children(payload);
      }}
    </Query>
  );
};

export default User;
