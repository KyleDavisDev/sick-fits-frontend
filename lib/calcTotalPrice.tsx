export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
