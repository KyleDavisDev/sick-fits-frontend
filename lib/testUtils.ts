import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
