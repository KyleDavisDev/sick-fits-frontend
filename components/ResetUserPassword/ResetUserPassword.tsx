import * as React from "react";
import gql from "graphql-tag";
import Form from "../Form/Form";
import ErrorMessage from "../ErrorMessage/ErrorMessage";
import { Mutation } from "react-apollo";
import { CURRENT_USER_QUERY } from "../User/User";

export const RESET_USER_PASSWORD_MUTATION = gql`
  mutation RESET_USER_PASSWORD_MUTATION(
    $resetToken: String!
    $password: String!
    $confirmPassword: String!
  ) {
    resetPassword(
      resetToken: $resetToken
      password: $password
      confirmPassword: $confirmPassword
    ) {
      email
      id
      name
    }
  }
`;

export interface IResetUserPasswordProps {
  resetToken: string;
}

export interface IResetUserPasswordState {
  resetToken: string;
  password: string;
  confirmPassword: string;
  [name: string]: string | boolean;
}

export default class ResetUserPassword extends React.Component<
  IResetUserPasswordProps,
  IResetUserPasswordState
> {
  state = {
    resetToken: this.props.resetToken,
    password: "",
    confirmPassword: ""
  };

  public render() {
    return (
      <Mutation
        mutation={RESET_USER_PASSWORD_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(resetPassword: any, { error, loading, called }: any) => {
          return (
            <Form
              onSubmit={async e => {
                e.preventDefault();
                await this.onFormSubmit(resetPassword);
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <ErrorMessage error={error} />
                {!loading && !error && called && (
                  <p>Success! Check your email for an update link</p>
                )}
                <h2>Reset your password</h2>

                <label htmlFor="password">
                  Password
                  <input
                    type="password"
                    name="password"
                    value={this.state.password}
                    onChange={this.saveToState}
                  ></input>
                </label>

                <label htmlFor="confirmPassword">
                  Confirm Password
                  <input
                    type="password"
                    name="confirmPassword"
                    value={this.state.confirmPassword}
                    onChange={this.saveToState}
                  ></input>
                </label>

                <button type="submit">Reset password!x</button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }

  private saveToState = (event: React.ChangeEvent<HTMLInputElement>) => {
    // grab name and value
    const name = event.target.name;
    const val = event.target.value;

    // set state
    this.setState({ [name]: val });
  };

  private onFormSubmit = async (resetPassword: any) => {
    // set loading state
    this.setState({ loading: true, error: null });

    await resetPassword();

    this.setState({
      password: "",
      confirmPassword: ""
    });
  };
}
