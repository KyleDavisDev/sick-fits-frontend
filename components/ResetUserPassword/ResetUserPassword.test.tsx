import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";
import ResetUserPassword, {
  RESET_USER_PASSWORD_MUTATION
} from "./ResetUserPassword";
import { CURRENT_USER_QUERY } from "../User/User";
import { fakeUser, simulateInputChange } from "../../lib/testUtils";

const mocks = [
  {
    request: {
      query: RESET_USER_PASSWORD_MUTATION,
      variables: { resetToken: "abc", password: "123", confirmPassword: "123" }
    },
    result: {
      data: {
        resetPassword: {
          __typename: "User",
          email: "abc@email.com",
          id: "abasdfasdf",
          name: "SampleName"
        }
      }
    }
  }
];

describe("<ResetUserPassword />", () => {
  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ResetUserPassword resetToken={"abc"} />
      </MockedProvider>
    );

    // check renders
    expect(wrapper).toBeTruthy();

    // check snapshot
    expect(wrapper.find("form")).toMatchSnapshot();
  });
});
