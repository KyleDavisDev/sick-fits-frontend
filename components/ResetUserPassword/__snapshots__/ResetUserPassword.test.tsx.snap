// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetUserPassword /> renders and matches snapshot 1`] = `
<form
  className="sc-bdVaJa cgcGkz"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <ErrorMessage
      error={Object {}}
    />
    <h2>
      Reset your password
    </h2>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="confirmPassword"
    >
      Confirm Password
      <input
        name="confirmPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Reset password!x
    </button>
  </fieldset>
</form>
`;
